nym-vpn-desktop/src-tauri/src/events.rs

```diff
@@ -0,0 +1,68 @@
+use tauri::Manager;
+use time::OffsetDateTime;
+use tracing::debug;
+
+use crate::states::app::ConnectionState;
+
+pub const EVENT_CONNECTION_STATE: &str = "connection-state";
+pub const EVENT_CONNECTION_PROGRESS: &str = "connection-progress";
+
+#[derive(Clone, serde::Serialize)]
+pub enum ConnectProgressMsg {
+    Initializing,
+    InitDone,
+}
+
+#[derive(Clone, serde::Serialize)]
+pub struct ProgressEventPayload {
+    pub key: ConnectProgressMsg,
+}
+
+#[derive(Clone, serde::Serialize)]
+pub struct ConnectionEventPayload {
+    state: ConnectionState,
+    error: Option<String>,
+    start_time: Option<i64>, // unix timestamp in seconds
+}
+
+impl ConnectionEventPayload {
+    pub fn new(state: ConnectionState, error: Option<String>, start_time: Option<i64>) -> Self {
+        Self {
+            state,
+            error,
+            start_time,
+        }
+    }
+}
+
+pub trait AppHandleEventEmitter {
+    fn emit_disconnected(&self, error: Option<String>);
+    fn emit_connected(&self, now: OffsetDateTime, gateway: String);
+}
```

Comment:
better if you also implement `emit_{disconnecting/connecting}`

Reply:
yep good point!